query/src/storages/result/download.rs

```diff
@@ -0,0 +1,93 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::str::FromStr;
+use std::sync::Arc;
+
+use async_stream::stream;
+use common_datablocks::DataBlock;
+use common_exception::Result;
+use common_io::prelude::FormatSettings;
+use common_planners::ReadDataSourcePlan;
+use common_planners::SourceInfo;
+use futures::StreamExt;
+
+use crate::servers::http::formats::tsv_output::block_to_tsv;
+use crate::sessions::QueryContext;
+use crate::storages::result::ResultTable;
+use crate::storages::Table;
+
+pub type SendableVu8Stream =
+    std::pin::Pin<Box<dyn futures::stream::Stream<Item = Result<Vec<u8>>> + Send>>;
+
+#[derive(Clone, Copy)]
+pub enum DownloadFormatType {
+    Tsv,
+}
+
+impl Default for DownloadFormatType {
+    fn default() -> Self {
+        Self::Tsv
+    }
+}
+
+impl FromStr for DownloadFormatType {
+    type Err = String;
+    fn from_str(s: &str) -> std::result::Result<Self, String> {
+        match s.to_uppercase().as_str() {
+            "TSV" => Ok(DownloadFormatType::Tsv),
+            _ => Err("Unknown file format type, must be one of { TSV }".to_string()),
+        }
+    }
+}
+
+fn serialize_block(block: DataBlock, fmt: DownloadFormatType) -> Result<Vec<u8>> {
```

Comment:
maybe it's time to move this util function into `query/src/formats/format_xxx.rs`

Reply:
ok

put input and output logics of `csv` in the same file `format_csv.rs`?  cc @zhang2014 
output interface is very diff from output.


other related:

- query::formats::format::{FormatFactory, InputFormat}
  -  input only
- comon_io::format_settings::FormatSettings 
  -  output only
- common::meta::types::user_stage{StageFileFormatType, FileFormatOptions}
  -  input/output only

do we need a unified enum for FileFormatType?

Reply:
In https://github.com/datafuselabs/databend/pull/5399, I moved the storage-related config into `common-io` and removed all stage's own storage config. I think it's OK to add a new FileFormatType here.

Only one note: `FileFormatType` is part of our contract to meta API, we should make it will not break `proto-conv`.